### PR TITLE
Use ArraySlice to slice a Vector

### PR DESCRIPTION
**Pull request by jdunkerley**
[ci no changelog needed]

### Pull Request Description

Use an `ArraySlice` to slice `Vector`.
Avoids memory copying for the slice function.

### Important Notes
| Test | Ref | New |
| --- | --- | --- |
| New Vector | 71.9 | 71.0 |
| Append Single | 26.0 | 27.7 |
| Append Large | 15.1 | 14.9 |
| Sum | 156.4 | 165.8 |
| Drop First 20 and Sum | 171.2 | 165.3 |
| Drop Last 20 and Sum | 170.7 | 163.0 |
| Filter | 76.9 | 76.9 |
| Filter With Index | 166.3 | 168.3 |
| Partition | 278.5 | 273.8 |
| Partition With Index | 392.0 | 393.7 |
| Each | 101.9 | 102.7 |

- Note: the performance of New and Append has got slower from previous tests.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso [hunk 1]
```diff
@@ -660,10 +660,10 @@ type Vector a
...
         if slice_start >= slice_end then [] else
             if (slice_start == 0) && (slice_end == self.length) then self else
-                len = slice_end - slice_start
-                arr = Array.new len
-                Array.copy self.to_array slice_start arr 0 len
-                Vector.from_polyglot_array arr
+                Vector.from_polyglot_array (self.slice_builtin slice_start slice_end)
+
+    ## PRIVATE
+    slice_builtin self start end = @Builtin_Method "Vector.slice_builtin"
```
Comment (JaroslavTulach) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso]: Interesting test: 
```
v = [1, 2, 3]
Meta.type_of v . should_equal Meta.type_of (v.slice 1 1)
```

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso [hunk 2]
```diff
@@ -660,10 +660,10 @@ type Vector a
         slice_end = Math.min self.length end
         if slice_start >= slice_end then [] else
             if (slice_start == 0) && (slice_end == self.length) then self else
-                len = slice_end - slice_start
-                arr = Array.new len
-                Array.copy self.to_array slice_start arr 0 len
-                Vector.from_polyglot_array arr
+                Vector.from_polyglot_array (self.slice_builtin slice_start slice_end)
```
Comment (JaroslavTulach) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso]: `Vector.from_polyglot_array` wraps the slice into yet another object.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+import org.enso.interpreter.node.expression.builtin.interop.syntax.HostValueToEnsoNode;
+
+@ExportLibrary(InteropLibrary.class)
+public class ArraySlice implements TruffleObject {
+  private final Object storage;
+  private final long start;
+  private final long length;
+
+  public ArraySlice(Object storage, long start, long length) {
+    if (storage instanceof ArraySlice slice) {
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java:21]: This condition is unlikely to be `true` - the code in https://github.com/enso-org/enso/pull/3724/files#r976005007 seems to wrap `ArraySlice` into yet another `Vector`.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso (hunk 1 above)
Comment (jdunkerley) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso]: no `type_of` yet so did similar tests.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+      this.storage = slice.storage;
+      this.start = slice.start + start;
+      this.length = length;
+    } else {
+      if (CompilerDirectives.inInterpreter()) {
+        if (!InteropLibrary.getUncached().hasArrayElements(storage)) {
+          throw new IllegalStateException("Vector needs array-like delegate, but got: " + storage);
+        }
+      }
```
Comment (radeusgd) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java]: I think I don't understand why this check is performed only in the interpreter. Won't this affect semantics depending on optimization levels?

**Review by hubertp — APPROVED** [review]

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Vector.java [hunk 5]
```diff
@@ -67,6 +63,26 @@ public final Object toArray() {
     return this.storage;
   }
 
+  @Builtin.Method(name = "slice", description = "Returns a slice of this Vector.")
+  @Builtin.Specialize
+  @Builtin.WrapException(from = UnsupportedMessageException.class, to = PanicException.class)
+  public final Vector slice(long start, long end, InteropLibrary interop)
+      throws UnsupportedMessageException {
```
Comment (radeusgd) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Vector.java:70]: In what circumstances is `UnsupportedMessageException` thrown?

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java (hunk 4 above)
Comment (radeusgd) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java]: Anyway, if left as-is I think it would be beneficial to add a short comment explaining why this is only checked in interpreter.

Unless there is some obvious reason that every Truffle developer should understand immediately, I think such an explanation will help future developers understand this decision.

File: test/Benchmarks/src/Vector.enso [hunk 6]
```diff
@@ -1,4 +1,5 @@
 from Standard.Base import all
+import Standard.Base
```
Comment (radeusgd) [inline test/Benchmarks/src/Vector.enso:2]: Why is this needed?

File: test/Tests/src/Data/Vector_Spec.enso [hunk 7]
```diff
@@ -212,6 +212,15 @@ spec = Test.group "Vectors" <|
         concat = [1, 2, 3] + [4, 5, 6]
         concat.should_equal [1, 2, 3, 4, 5, 6]
 
+    Test.specify "Vector slice should return a Vector" <|
+        vec = [1, 2, 3, 4, 5, 6]
+        vec.slice 0 3 . should_equal [1, 2, 3]
+        vec.slice 1 3 . should_equal [2, 3]
+        vec.slice 1 1 . should_equal []
+        vec.slice 0 100 . should_equal [1, 2, 3, 4, 5, 6]
```
Comment (radeusgd) [inline test/Tests/src/Data/Vector_Spec.enso:220]:
```suggestion
        vec.slice 0 100 . should_equal vec
```

**Review by radeusgd — APPROVED** [review]
Looks good to me overall, just a few small comments and a question on implementation that would be good if can be answered before merge.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java (hunk 4 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java]: We could do the check as `assert`. Asserts are ignored when Truffle does its PE compilation. We just need to make sure we run our tests with `-ea`.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+  @ExportMessage
+  public boolean hasArrayElements() {
+    return true;
+  }
+
+  @ExportMessage
+  public long getArraySize(@CachedLibrary(limit = "3") InteropLibrary interop)
+      throws UnsupportedMessageException {
+    long storageSize = interop.getArraySize(storage);
+    return Math.min(storageSize, end) - start;
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java]: This could get negative, right?

Creates slice over JavaScript array and then remove something from the underlying array. The `storageSize` gets smaller and substracting `start` may give us negative index.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
+package org.enso.interpreter.runtime.data;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.interop.*;
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java]: I'd replace the star import with individual classes imports in the name of clarity.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+import com.oracle.truffle.api.library.ExportMessage;
+import org.enso.interpreter.node.expression.builtin.interop.syntax.HostValueToEnsoNode;
+
+@ExportLibrary(InteropLibrary.class)
+public class ArraySlice implements TruffleObject {
+  private final Object storage;
+  private final long start;
+  private final long end;
+
+  public ArraySlice(Object storage, long start, long end) {
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java]: Who's supposed to call this constructor? Just `Vector.slice` from the same package? Then make it package private. Possibly also make the whole class package private.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Vector.java [hunk 11]
```diff
@@ -67,6 +63,26 @@ public final Object toArray() {
     return this.storage;
   }
 
+  @Builtin.Method(name = "slice", description = "Returns a slice of this Vector.")
+  @Builtin.Specialize
+  @Builtin.WrapException(from = UnsupportedMessageException.class, to = PanicException.class)
+  public final Vector slice(long start, long end, InteropLibrary interop)
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Vector.java:69]: Is the `InteropLibrary` cached or uncached?

File: test/Tests/src/Data/Vector_Spec.enso (hunk 7 above)
Comment (JaroslavTulach) [reply test/Tests/src/Data/Vector_Spec.enso:220]: Radek, do you suggest to test for identity? Then the check has to use `Meta.is_same_object`. Otherwise `==` on _Arrays&Vectors_ always only compares content.

**Review by JaroslavTulach — APPROVED** [review]
Not bad. If that works, it can go in. There is `VectorBenchmarks.java` - consider adding a `sliced` check in there.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Vector.java (hunk 11 above)
Comment (hubertp) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Vector.java:69]: If you look at the generated node it is cached

File: test/Tests/src/Data/Vector_Spec.enso (hunk 7 above)
Comment (radeusgd) [reply test/Tests/src/Data/Vector_Spec.enso:220]: No, I suggest the check on content, I just don't see a reason for writing `[1, 2, 3, 4, 5, 6]` twice if its exactly the content of `vec`. But I get your point that this could _falsely_ suggest a check on _identity_ so probably good to leave it as-is.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java (hunk 4 above)
Comment (radeusgd) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArraySlice.java]: Why can't the check be there _always_?

File: test/Benchmarks/src/Vector.enso (hunk 6 above)
Comment (jdunkerley) [reply test/Benchmarks/src/Vector.enso:2]: As the Benchmark is called `Vector` means you can access the `Standard.Base.Vector` as `Base.Vector`.

Comment (radeusgd) [reply test/Benchmarks/src/Vector.enso:2]: Ah, okay. I was confused as there weren't any other changes.

But I assume before `from Standard.Base import all` was introducing `Base` into scope and this changed in one of the recent PRs and the benchmark was just not tested since then. So now it makes sense, thanks for the explanation.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Vector.java (hunk 5 above)
Comment (jdunkerley) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Vector.java:70]: Only if the backing storage no longer can answer the `interop.getArraySize` call - shouldn't happen.
